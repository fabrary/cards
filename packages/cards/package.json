{
  "name": "@flesh-and-blood/cards",
  "description": "TypeScript representations of Flesh and Blood cards",
  "version": "3.6.96",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "sideEffects": false,
  "scripts": {
    "transform": "ts-node scripts/index.ts",
    "prebuild": "rm -rf dist",
    "build": "esbuild ./src/index.ts --outdir=dist --bundle --minify --platform=node --target=es6",
    "postbuild": "tsc --declaration && npm test",
    "test": "jest",
    "update:dependencies": "ncu -u latest-cards && npm i",
    "update:snapshots": "npm test -- -u",
    "new": "ts-node tests/new-cards.ts",
    "tts": "ts-node tts/index.ts",
    "sitemap": "ts-node sitemap/index.ts",
    "full": "npm run transform && npm run build && npm test && npm run new && npm run tts && npm run sitemap",
    "prepublish": "npm run full"
  },
  "dependencies": {
    "@flesh-and-blood/types": "^3.6.96"
  },
  "devDependencies": {
    "@types/jest": "^29.5.11",
    "@types/node": "^20.10.5",
    "@types/papaparse": "5.3.14",
    "chalk": "^4.1.2",
    "esbuild": "^0.19.10",
    "esbuild-jest": "^0.5.0",
    "fast-xml-parser": "4.3.2",
    "jest": "^29.7.0",
    "latest-cards": "npm:@flesh-and-blood/cards@^3.6.96",
    "papaparse": "^5.3.2",
    "ts-node": "^10.9.2",
    "typescript": "^5.3.3",
    "typo-js": "^1.2.3"
  },
  "files": [
    "/dist"
  ],
  "publishConfig": {
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/fabrary/fab-cards.git"
  },
  "bugs": {
    "url": "https://github.com/fabrary/fab-cards/issues"
  },
  "homepage": "https://github.com/fabrary/fab-cards#readme",
  "author": "Phillip Manwaring",
  "license": "MIT",
  "keywords": [
    "Flesh and Blood",
    "FAB",
    "FABTCG"
  ],
  "gitHead": "71162394e6b8ac0ef9a01cdaacc2661d4a34ae3e"
}
