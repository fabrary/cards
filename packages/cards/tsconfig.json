{
  "compilerOptions": {
    "emitDeclarationOnly": true,
    "esModuleInterop": true,
    "declaration": true,
    "isolatedModules": true,
    "lib": ["ES2021"],
    "module": "NodeNext",
    "moduleResolution": "nodenext",
    "outDir": "dist",
    "skipLibCheck": true,
    "target": "ES6"
  },
  "include": ["src"]
}
