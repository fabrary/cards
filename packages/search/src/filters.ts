import {
  Card,
  Foiling,
  Hero,
  Meta,
  Release,
  Treatment,
  Type,
  setIdentifierToSetMappings,
  setToSetIdentifierMappings,
} from "@flesh-and-blood/types";
import { getAbbreviation } from "./abbreviations";
import { getExcludedMetaFilters, getMetaFilters } from "./metaFilters";
import {
  multiWordShorthands,
  shorthands,
  singleWordShorthands,
} from "./shorthands";
import { PUNCTUATION } from "./constants";
import { getCardByName, getRelatedCardsByName } from ".";

export interface AppliedFilter {
  filterToPropertyMapping: FilterToPropertyMapping;
  values: string[];
  isAnd?: boolean;
  isOr?: boolean;
  modifier?: Modifier;
  excluded?: boolean;
  cardTypes?: string[];
}

export type Filter =
  | "art"
  | "artist"
  | "attack"
  | "banned"
  | "b"
  | "block"
  | "c"
  | "class"
  | "chain"
  | "co"
  | "cost"
  | "color"
  | "d"
  | "def"
  | "defense"
  | "f"
  | "fusion"
  | "foil"
  | "foiling"
  | "hero"
  | "i"
  | "intellect"
  | "k"
  | "keyword"
  | "l"
  | "legal"
  | "li"
  | "life"
  | "name"
  | "p"
  | "pitch"
  | "pwr"
  | "pow"
  | "power"
  | "r"
  | "rarity"
  | "referencedby"
  | "references"
  | "s"
  | "set"
  | "sp"
  | "specialization"
  | "specializations"
  | "st"
  | "subtype"
  | "text"
  | "t"
  | "type"
  | "tal"
  | "talent"
  | "text"
  | "treatment"
  | "variation"
  | "x";

export type Modifier = ">=" | ">" | "<=" | "<";
export const availableModifiers: Modifier[] = [">=", ">", "<=", "<"];

export type Exclusion = "!" | "-";
export const availableExclusions: Exclusion[] = ["!", "-"];

export interface FilterToPropertyMapping {
  nestedProperty?: string;
  property: string;
  exclusion?: Exclusion;
  isArray?: boolean;
  isNumber?: boolean;
  isString?: boolean;
  isBoolean?: boolean;
  isMeta?: boolean;
  modifier?: Modifier;
  partialMatch?: boolean;
  specialProperty?: string;
}

const arcaneFilter: FilterToPropertyMapping = {
  property: "arcane",
  specialProperty: "specialArcane",
  isNumber: true,
  partialMatch: true,
};

const artistFilter: FilterToPropertyMapping = {
  property: "artists",
  isArray: true,
  partialMatch: true,
};

const bannedFilter: FilterToPropertyMapping = {
  property: "bannedFormats",
  isArray: true,
  partialMatch: true,
};

const bondFilter: FilterToPropertyMapping = {
  property: "bonds",
  isArray: true,
};

const chainFilter: FilterToPropertyMapping = {
  property: "n/a",
};

const classFilter: FilterToPropertyMapping = {
  property: "classes",
  isArray: true,
  partialMatch: true,
};

const costFilter: FilterToPropertyMapping = {
  property: "cost",
  specialProperty: "specialCost",
  isNumber: true,
  partialMatch: true,
};

const defenseFilter: FilterToPropertyMapping = {
  property: "defense",
  specialProperty: "specialDefense",
  isNumber: true,
};

const flowFilter: FilterToPropertyMapping = {
  property: "flows",
  isArray: true,
};

const foilFilter: FilterToPropertyMapping = {
  nestedProperty: "foiling",
  property: "printings",
  isArray: true,
};

const fuseFilter: FilterToPropertyMapping = {
  property: "fusions",
  isArray: true,
};

const intellectFilter: FilterToPropertyMapping = {
  property: "intellect",
  isNumber: true,
};

const keywordFilter: FilterToPropertyMapping = {
  property: "keywords",
  isArray: true,
  partialMatch: true,
};

const legalFilter: FilterToPropertyMapping = {
  property: "n/a",
  isMeta: true,
};

const lifeFilter: FilterToPropertyMapping = {
  property: "life",
  specialProperty: "specialLife",
  isNumber: true,
};

const metaFilter: FilterToPropertyMapping = {
  property: "meta",
  isArray: true,
};

const nameFilter: FilterToPropertyMapping = {
  property: "name",
  isString: true,
  partialMatch: true,
};

const pitchFilter: FilterToPropertyMapping = {
  property: "pitch",
  isNumber: true,
};

const powerFilter: FilterToPropertyMapping = {
  property: "power",
  specialProperty: "specialPower",
  isNumber: true,
};

const rarityFilter: FilterToPropertyMapping = {
  property: "n/a",
  isMeta: true,
};

const referencedByFilter: FilterToPropertyMapping = {
  property: "n/a",
};

const referencesFilter: FilterToPropertyMapping = {
  property: "n/a",
};

const setFilter: FilterToPropertyMapping = {
  property: "sets",
  isArray: true,
  partialMatch: true,
};

const specializationsFilter: FilterToPropertyMapping = {
  property: "specializations",
  isArray: true,
  partialMatch: true,
};

const subtypeFilter: FilterToPropertyMapping = {
  property: "subtypes",
  isArray: true,
};

const typeFilter: FilterToPropertyMapping = {
  property: "types",
  isArray: true,
};

const talentFilter: FilterToPropertyMapping = {
  property: "talents",
  isArray: true,
};

const textFilter: FilterToPropertyMapping = {
  property: "functionalText",
  isString: true,
  partialMatch: true,
};

const traitFilter: FilterToPropertyMapping = {
  property: "traits",
  isArray: true,
  partialMatch: true,
};

const typeTextFilter: FilterToPropertyMapping = {
  property: "typeText",
  isString: true,
  partialMatch: true,
};

const treatmentFilter: FilterToPropertyMapping = {
  nestedProperty: "treatment",
  property: "printings",
  isArray: true,
};

export const filtersToCardPropertyMappings = {
  arcane: arcaneFilter,
  a: artistFilter,
  artist: artistFilter,
  art: artistFilter,
  attack: powerFilter,
  b: defenseFilter,
  block: defenseFilter,
  banned: bannedFilter,
  bond: bondFilter,
  bonds: bondFilter,
  c: classFilter,
  class: classFilter,
  chain: chainFilter,
  co: costFilter,
  cost: costFilter,
  color: pitchFilter,
  d: defenseFilter,
  def: defenseFilter,
  defense: defenseFilter,
  flow: flowFilter,
  flows: flowFilter,
  f: fuseFilter,
  fusion: fuseFilter,
  foil: foilFilter,
  foiling: foilFilter,
  i: intellectFilter,
  intellect: intellectFilter,
  is: metaFilter,
  k: keywordFilter,
  keyword: keywordFilter,
  l: legalFilter,
  legal: legalFilter,
  hero: legalFilter,
  li: lifeFilter,
  life: lifeFilter,
  meta: metaFilter,
  n: nameFilter,
  name: nameFilter,
  p: pitchFilter,
  pitch: pitchFilter,
  pwr: powerFilter,
  pow: powerFilter,
  power: powerFilter,
  r: rarityFilter,
  rarity: rarityFilter,
  referencedby: referencedByFilter,
  references: referencesFilter,
  rf: bannedFilter,
  s: setFilter,
  set: setFilter,
  sp: specializationsFilter,
  spec: specializationsFilter,
  specialization: specializationsFilter,
  specializations: specializationsFilter,
  st: subtypeFilter,
  subtype: subtypeFilter,
  t: typeFilter,
  type: typeFilter,
  tal: talentFilter,
  talent: talentFilter,
  text: textFilter,
  trait: traitFilter,
  treat: treatmentFilter,
  treatment: treatmentFilter,
  var: treatmentFilter,
  variation: treatmentFilter,
  x: typeTextFilter,
};

const punctuationOverrides = [
  {
    text: Release.ClassicBattlesRhinarDorinthea.toLowerCase(),
    override: Release.ClassicBattlesRhinarDorinthea.toLowerCase().replaceAll(
      PUNCTUATION,
      ""
    ),
  },
];
const getSearchCriteria = (text: string): string[] => {
  const searchCriteria: string[] = [];
  let rawSearchCriteria = text.replaceAll("‚Äù", '"');
  for (const { text, override } of punctuationOverrides) {
    if (rawSearchCriteria.includes(text)) {
      rawSearchCriteria = rawSearchCriteria.replace(text, override);
    }
  }

  const matchingAbbreviation = getAbbreviation(rawSearchCriteria)?.card;
  if (matchingAbbreviation) {
    searchCriteria.push(rawSearchCriteria);
  } else {
    const splitSearchCriteria = rawSearchCriteria.split(/[ ]+/);
    let combinedSearchCriteria = "";
    let combinedQuoteCount = 0;
    for (const criteria of splitSearchCriteria) {
      if (combinedQuoteCount === 2) {
        searchCriteria.push(combinedSearchCriteria.trim().replaceAll('"', ""));
        combinedSearchCriteria = "";
        combinedQuoteCount = 0;
      }
      if (combinedQuoteCount < 2 && criteria.split('"').length === 2) {
        combinedSearchCriteria += " " + criteria;
        combinedQuoteCount++;
      } else if (combinedQuoteCount === 0 && combinedSearchCriteria === "") {
        searchCriteria.push(criteria);
      } else if (combinedQuoteCount === 1) {
        combinedSearchCriteria += " " + criteria;
      }
    }
    if (combinedQuoteCount === 2) {
      searchCriteria.push(combinedSearchCriteria.trim().replaceAll('"', ""));
      combinedSearchCriteria = "";
      combinedQuoteCount = 0;
    }
  }
  return searchCriteria;
};

export interface SpecialConditions {
  isClash: boolean;
  heroes: Hero[];
}

const specialConditionHeroes = [
  Hero.Briar,
  Hero.Brutus,
  Hero.Emperor,
  Hero.Prism,
  Hero.Shiyana,
  Hero.Taylor,
  Hero.Yorick,
];

export const getKeywordsAndAppliedFiltersFromText = (
  text: string,
  cards: Card[]
): {
  appliedFilters: AppliedFilter[];
  attributes: {
    artists: string[];
    foilings: Foiling[];
    isExpansionSlot: boolean;
    releases: Release[];
    treatments: Treatment[];
  };
  keywords: string[];
  specialConditions?: SpecialConditions;
} => {
  let expandedText = text.trim().toLowerCase();
  for (const { filters, shorthands } of multiWordShorthands) {
    for (const shorthand of shorthands) {
      if (expandedText.includes(shorthand)) {
        expandedText = expandedText.replace(shorthand, filters.join(" "));
        break;
      }
    }
  }
  for (const [set, setIdentifiers] of Object.entries(
    setToSetIdentifierMappings
  )) {
    if (expandedText.includes(set.toLowerCase())) {
      expandedText = expandedText.replace(set.toLowerCase(), setIdentifiers[0]);
    }
  }
  const rawSearchCriteria = getSearchCriteria(expandedText);

  const searchCriteria: string[] = [];
  for (const criteria of rawSearchCriteria) {
    const expanded = singleWordShorthands.find(({ shorthands }) =>
      shorthands.includes(criteria)
    );
    if (expanded) {
      searchCriteria.push(...expanded.filters);
    } else {
      searchCriteria.push(criteria);
    }
  }

  let appliedFilters: AppliedFilter[] = [];
  let artists: string[] = [];
  let keywords: string[] = [];
  let foilings: Foiling[] = [];
  let isExpansionSlot: boolean = false;
  let releases: Release[] = [];
  let treatments: Treatment[] = [];
  let specialConditions: SpecialConditions = { isClash: false, heroes: [] };

  for (const criteria of searchCriteria) {
    if (hasFilter(criteria)) {
      const [unparsedFilterKey, unparsedFilterValue] = criteria.split(":");

      let { modifier, values, isAnd, isOr } =
        getFilterValuesAndModifier(unparsedFilterValue);
      let { filterKey, excluded, isMeta } =
        getFilterKeyAndExcluded(unparsedFilterKey);

      if (isMeta) {
        if (["rarity", "r"].includes(filterKey)) {
          values = getRarityValuesFromText(values);
        }
        if (["legal", "l", "hero"].includes(filterKey)) {
          for (const hero of specialConditionHeroes) {
            // if (text.toLowerCase() === "legal:shiyana figment") {
            //   console.log({
            //     hero,
            //     specialConditions,
            //     unparsedFilterValue,
            //     includes: unparsedFilterValue.includes(hero.toLowerCase()),
            //     alreadyAdded: specialConditions.heroes.includes(hero),
            //   });
            // }
            if (unparsedFilterValue.includes(hero.toLowerCase())) {
              if (!specialConditions.heroes.includes(hero)) {
                specialConditions.heroes.push(hero);
              }
            }
          }
          if (unparsedFilterValue.includes("clash")) {
            specialConditions.isClash = true;
          }
        }
        appliedFilters.push(
          ...getMetaFilters(excluded, filterKey, values, modifier)
        );
      } else {
        if (["chain"].includes(filterKey)) {
          const getName = (card: Card) =>
            card.name.toLowerCase().replaceAll(PUNCTUATION, "");

          const relatedCardNames: string[] = values
            .map((name) => getCardByName(name, cards))
            .filter((card) => !!card)
            .map(getName);

          const names = new Set<string>(relatedCardNames);

          filterKey = "name";
          isOr = true;
          let limit = 20;
          let counter = 0;

          const addToSetAndRelated = (card: Card) => {
            if (!card.types.includes(Type.Hero)) {
              const name = getName(card);
              names.add(name);
              if (!relatedCardNames.includes(name)) {
                relatedCardNames.push(name);
              }
            }
          };

          for (const relatedCardName of relatedCardNames) {
            if (counter > limit) {
              break;
            }
            const { referencedBy, references } = getRelatedCardsByName(
              relatedCardName,
              cards
            );
            references.forEach(addToSetAndRelated);
            if (counter === 0) {
              referencedBy.forEach(addToSetAndRelated);
            }
            counter++;
          }
          values = Array.from(names);
        } else if (["referencedby", "references"].includes(filterKey)) {
          const originalKey = filterKey;
          const relatedCardNames = [...values].filter((value) => !!value);
          values = [];
          filterKey = "name";
          isOr = true;
          for (const relatedCardName of relatedCardNames) {
            const { referencedBy, references } = getRelatedCardsByName(
              relatedCardName,
              cards
            );
            if (["referencedby"].includes(originalKey)) {
              values.push(
                ...references.map(({ name }) =>
                  name.toLowerCase().replaceAll(PUNCTUATION, "")
                )
              );
            } else if (["references"].includes(originalKey)) {
              values.push(
                ...referencedBy.map(({ name }) =>
                  name.toLowerCase().replaceAll(PUNCTUATION, "")
                )
              );
            }
          }
        } else if (["art", "artist"].includes(filterKey)) {
          artists = values;
        } else if (["is", "is", "meta"].includes(filterKey)) {
          const metaValues = getMetaValuesFromText(values);
          values = metaValues.map((v) =>
            v.toLowerCase().replaceAll(PUNCTUATION, "")
          );
          if (metaValues.includes(Meta.Expansion) && !excluded) {
            isExpansionSlot = true;
          }
        } else if (["foiling", "foil"].includes(filterKey)) {
          foilings = getFoilingValuesFromText(values);
          values = foilings.map((f) => f.toLowerCase());
        } else if (
          ["treat", "treatment", "var", "variation"].includes(filterKey)
        ) {
          treatments = getTreatmentValuesFromText(values);
          values = treatments.map((t) => t.toLowerCase());
        } else if (["set", "s"].includes(filterKey)) {
          releases = getReleasesFromRawValues(values);
          values = releases.map((s) =>
            s.toLowerCase().replaceAll(PUNCTUATION, "")
          );
        } else if (["pitch", "p", "color"].includes(filterKey)) {
          values = getPitchValuesFromText(values);
        }
        if (filtersToCardPropertyMappings[filterKey as Filter]) {
          appliedFilters.push({
            filterToPropertyMapping:
              filtersToCardPropertyMappings[filterKey as Filter],
            values,
            isAnd,
            isOr,
            modifier,
            excluded,
          });
        }
      }
    } else if (criteria) {
      const expanded = getAbbreviation(criteria)?.card;
      const metaFilters = getExcludedMetaFilters(criteria);
      if (expanded) {
        keywords.push(`"${expanded.toLowerCase().replace(PUNCTUATION, "")}"`);
      } else if (metaFilters && metaFilters.length > 0) {
        appliedFilters.push(...metaFilters);
      } else {
        keywords.push(criteria.replace(PUNCTUATION, ""));
      }
    }
  }
  return {
    appliedFilters,
    attributes: { artists, foilings, isExpansionSlot, releases, treatments },
    keywords,
    specialConditions,
  };
};

const getReleasesFromRawValues = (rawValues: string[]): Release[] => {
  const releases: Release[] = [];
  for (const rawValue of rawValues) {
    releases.push(...getMatchingReleasesFromRawValue(rawValue));
  }
  return releases;
};

const getMatchingReleasesFromRawValue = (rawValue: string) => {
  const releases: Release[] = [];

  const setFromValue = Object.values(Release).find(
    (release) => release.toLowerCase().replaceAll(PUNCTUATION, "") === rawValue
  );

  if (setFromValue) {
    releases.push(setFromValue);
  }

  if (releases.length === 0) {
    const setFromSetIdentifier = setIdentifierToSetMappings[rawValue];
    if (setFromSetIdentifier) {
      releases.push(setFromSetIdentifier);
    }
  }

  if (releases.length === 0) {
    const setsFromPartialValue = Object.values(Release).filter((release) =>
      release.toLowerCase().includes(rawValue)
    );
    if (setsFromPartialValue.length > 0) {
      releases.push(...setsFromPartialValue);
    }
  }

  return releases;
};

const pitchValuesMapping: { [key: string]: number } = {
  blue: 3,
  yellow: 2,
  red: 1,
  white: 0,
};
const getPitchValuesFromText = (rawValues: string[]) => {
  const values: string[] = [];
  for (const rawValue of rawValues) {
    if (pitchValuesMapping[rawValue] || pitchValuesMapping[rawValue] === 0) {
      values.push(pitchValuesMapping[rawValue].toString());
    } else {
      values.push(rawValue);
    }
  }
  return values;
};

const metaValuesMapping: { [key: string]: Meta } = {
  exp: Meta.Expansion,
  expansion: Meta.Expansion,
  expansionSlot: Meta.Expansion,
  rainbow: Meta.Rainbow,
};
const getMetaValuesFromText = (rawValues: string[]) => {
  const values: Meta[] = [];
  for (const rawValue of rawValues) {
    if (metaValuesMapping[rawValue]) {
      values.push(metaValuesMapping[rawValue]);
    }
  }
  return values;
};

const foilingValuesMapping: { [key: string]: Foiling } = {
  r: Foiling.R,
  rf: Foiling.R,
  rainbow: Foiling.R,
  c: Foiling.C,
  cf: Foiling.C,
  cold: Foiling.C,
  g: Foiling.G,
  gf: Foiling.G,
  gold: Foiling.G,
};
const getFoilingValuesFromText = (rawValues: string[]) => {
  const values: Foiling[] = [];
  for (const rawValue of rawValues) {
    if (foilingValuesMapping[rawValue]) {
      values.push(foilingValuesMapping[rawValue]);
    }
  }
  return values;
};

const treatmentValuesMapping: { [key: string]: Treatment } = {
  ...Object.values(Treatment).reduce((obj, treatment) => {
    obj[treatment.toLowerCase()] = treatment;
    return obj;
  }, {}),
  ...{
    aa: Treatment.AA,
    alt: Treatment.AA,
    "alt art": Treatment.AA,
    ab: Treatment.AB,
    "alt border": Treatment.AB,
    at: Treatment.AT,
    "alt text": Treatment.AT,
    // ds: Treatment.DS,
    // "double sided": Treatment.DS,
    ea: Treatment.EA,
    extended: Treatment.EA,
    "extended art": Treatment.EA,
    fa: Treatment.FA,
    full: Treatment.FA,
    "full art": Treatment.FA,
  },
};
const getTreatmentValuesFromText = (rawValues: string[]) => {
  const values: Treatment[] = [];
  for (const rawValue of rawValues) {
    if (treatmentValuesMapping[rawValue]) {
      values.push(treatmentValuesMapping[rawValue]);
    } else if (!!Treatment[rawValue.toUpperCase()]) {
      values.push(Treatment[rawValue.toUpperCase()]);
    }
  }
  return values;
};

const rarityValuesMapping: { [key: string]: string } = {
  c: "common",
  f: "fabled",
  l: "legendary",
  m: "majestic",
  v: "marvel",
  r: "rare",
  s: "super rare",
  t: "token",
};
const getRarityValuesFromText = (rawValues: string[]) => {
  const values: string[] = [];
  for (const rawValue of rawValues) {
    if (rarityValuesMapping[rawValue]) {
      values.push(rarityValuesMapping[rawValue]);
    } else {
      values.push(rawValue);
    }
  }
  return values;
};

const getFilterValuesAndModifier = (
  unparsedFilterValue: string
): {
  modifier: Modifier;
  values: string[];
  isAnd?: boolean;
  isOr?: boolean;
} => {
  const values: string[] = [];
  let isAnd, isOr;
  const modifier = availableModifiers.find((modifier) =>
    unparsedFilterValue.includes(modifier)
  ) as Modifier;
  if (modifier) {
    // if there's a modifier in the string, pull it out to get just the filter value
    const [, rawValue] = unparsedFilterValue.split(modifier);
    if (filterIsAnd(rawValue)) {
      isAnd = true;
      values.push(
        ...rawValue
          .trim()
          .split("+")
          .map((value) => value.replace(PUNCTUATION, ""))
      );
    } else if (filterIsOr(rawValue)) {
      isOr = true;
      values.push(
        ...rawValue
          .trim()
          .split(",")
          .map((value) => value.replace(PUNCTUATION, ""))
      );
    } else {
      values.push(rawValue.trim().replace(PUNCTUATION, ""));
    }
  } else {
    // otherwise the whole string is the filter value
    if (filterIsAnd(unparsedFilterValue)) {
      isAnd = true;
      values.push(
        ...unparsedFilterValue
          .trim()
          .split("+")
          .map((value) => value.replace(PUNCTUATION, ""))
      );
    } else if (filterIsOr(unparsedFilterValue)) {
      isOr = true;
      values.push(
        ...unparsedFilterValue
          .trim()
          .split(",")
          .map((value) => value.replace(PUNCTUATION, ""))
      );
    } else {
      if (
        unparsedFilterValue.startsWith('"') &&
        unparsedFilterValue.endsWith('"')
      ) {
        values.push(
          unparsedFilterValue
            .trim()
            .replaceAll('"', "")
            .replace(PUNCTUATION, "")
        );
      } else {
        values.push(unparsedFilterValue.trim().replace(PUNCTUATION, ""));
      }
    }
  }

  return { modifier, values, isAnd, isOr };
};

const getFilterKeyAndExcluded = (
  unparsedFilterKey: string
): {
  filterKey: string;
  excluded: boolean;
  isMeta?: boolean;
} => {
  const exclusion = getExclusion(unparsedFilterKey);
  if (exclusion) {
    const [, filterKey] = unparsedFilterKey.split(exclusion);
    return { excluded: true, filterKey, isMeta: filterIsMeta(filterKey) };
  } else {
    return {
      excluded: false,
      filterKey: unparsedFilterKey,
      isMeta: filterIsMeta(unparsedFilterKey),
    };
  }
};

const hasFilter = (text: string): boolean => text.indexOf(":") >= 0;
const filterIsAnd = (text: string): boolean => text.indexOf("+") >= 0;
const filterIsOr = (text: string): boolean => text.indexOf(",") >= 0;
const filterIsMeta = (filterKey: string): boolean =>
  !!filtersToCardPropertyMappings[filterKey as Filter]?.isMeta;
const getExclusion = (text: string): Exclusion =>
  availableExclusions
    .find((exclusion) => text.includes(exclusion))
    ?.slice(0, 1) as Exclusion;
