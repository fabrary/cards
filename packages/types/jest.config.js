module.exports = {
  expand: true,
  // silent: true,
  testEnvironment: "node",
  testMatch: ["<rootDir>/tests/**.test.ts"],
  transform: {
    "^.+\\.tsx?$": "esbuild-jest",
  },
};
