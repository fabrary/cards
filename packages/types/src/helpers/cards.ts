import { Type, Keyword, Subtype } from "../interfaces";

export const getCardIdentifier = (
  card: {
    name: string;
    pitch?: string | number;
  },
  useNumber?: boolean
): string => {
  const { name: unformattedName, pitch } = card;
  const name = unformattedName
    .toLowerCase()
    .replace("//", "-")
    .normalize("NFD")
    .replace(/\p{Diacritic}/gu, "")
    .replace(/ /g, "-")
    .replace("Ã­", "i")
    .replace(/[^a-z0-9 -]/g, "")
    .replace(/--/, "-");

  let suffix: string = "";
  switch (pitch) {
    case "1":
    case 1:
      suffix = useNumber ? "-1" : "-red";
      break;
    case "2":
    case 2:
      suffix = useNumber ? "-2" : "-yellow";
      break;
    case "3":
    case 3:
      suffix = useNumber ? "-3" : "-blue";
      break;
  }
  return `${name}${suffix}`;
};

export const getIsArenaCard = (types: Type[], keywords?: Keyword[]) => {
  const isDeckCard = getIsDeckCard(types, keywords);

  const isInventoryCard = [Type.DemiHero, Type.Equipment, Type.Weapon].some(
    (type) => types.includes(type)
  );

  return !isDeckCard && isInventoryCard;
};

export const getIsDeckCard = (types: Type[], keywords?: Keyword[]) => {
  const isDeckCardType = [
    Type.Action,
    Type.AttackReaction,
    Type.Block,
    Type.DefenseReaction,
    Type.Instant,
    Type.Mentor,
    Type.Resource,
  ].some((type) => types.includes(type));

  const isNotEphemeral = !keywords || !keywords.includes(Keyword.Ephemeral);

  return isDeckCardType && isNotEphemeral;
};

export const getCanAddToDeck = (types: Type[], keywords?: Keyword[]) => {
  const isArenaCard = getIsArenaCard(types, keywords);
  const isDeckCard = getIsDeckCard(types, keywords);

  return isArenaCard || isDeckCard;
};

export const getShouldRotateCardImage = (card: {
  keywords?: Keyword[];
  subtypes?: Subtype[];
}): boolean =>
  card.keywords?.includes(Keyword.Meld) ||
  card.subtypes?.includes(Subtype.Landmark) ||
  false;
