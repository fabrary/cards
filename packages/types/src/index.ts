export * from "./interfaces";
export * from "./sets";
