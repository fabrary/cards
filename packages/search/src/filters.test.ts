import { Foiling, Meta, Release, Treatment } from "@flesh-and-blood/types";
import { getKeywordsAndAppliedFiltersFromText } from "./filters";
import { cards } from "@flesh-and-blood/cards";

describe("Gets the right attribute filters", () => {
  const foilFilters = [
    ["foiling:r", [Foiling.R]],
    ["foil:cf,g", [Foiling.C, Foiling.G]],
  ];
  it.each(foilFilters)(
    "Gets matching foilings for %s",
    (search, expectedFoilings) => {
      const {
        attributes: { foilings },
      } = getKeywordsAndAppliedFiltersFromText(search as string, cards);
      expect(foilings.length).toEqual(expectedFoilings.length);
      expect(expectedFoilings).toMatchSnapshot();
    }
  );

  const setFilters = [
    ["s:ara", [Release.ArakniBlitzDeck]],
    ["s:welcome", [Release.WelcomeToRathe, Release.IraWelcomeDeck]],
  ];
  it.each(setFilters)("Gets matching sets for %s", (search, sets) => {
    const {
      attributes: { releases },
    } = getKeywordsAndAppliedFiltersFromText(search as string, cards);
    expect(releases.length).toEqual(sets.length);
    expect(releases).toMatchSnapshot();
  });

  const treatmentFilters = [
    ["treatment:ea", [Treatment.EA]],
    ['treat:"alt art",fa', [Treatment.AA, Treatment.FA]],
  ];
  it.each(treatmentFilters)(
    "Gets matching foilings for %s",
    (search, expectedTreatments) => {
      const {
        attributes: { treatments },
      } = getKeywordsAndAppliedFiltersFromText(search as string, cards);
      expect(treatments.length).toEqual(expectedTreatments.length);
      expect(treatments).toMatchSnapshot();
    }
  );
});

describe("Gets the right attribute filters", () => {
  const foilFilters = [
    ["foiling:r", [Foiling.R]],
    ["foil:cf,g", [Foiling.C, Foiling.G]],
  ];
  it.each(foilFilters)(
    "Gets matching foilings for %s",
    (search, expectedFoilings) => {
      const {
        attributes: { foilings },
      } = getKeywordsAndAppliedFiltersFromText(search as string, cards);
      expect(foilings.length).toEqual(expectedFoilings.length);
      expect(expectedFoilings).toMatchSnapshot();
    }
  );

  const setFilters = [
    ["s:ara", [Release.ArakniBlitzDeck]],
    ["s:welcome", [Release.WelcomeToRathe, Release.IraWelcomeDeck]],
  ];
  it.each(setFilters)("Gets matching sets for %s", (search, sets) => {
    const {
      attributes: { releases },
    } = getKeywordsAndAppliedFiltersFromText(search as string, cards);
    expect(releases.length).toEqual(sets.length);
    expect(releases).toMatchSnapshot();
  });

  const treatmentFilters = [
    ["treatment:ea", [Treatment.EA]],
    ['treat:"alt art",fa', [Treatment.AA, Treatment.FA]],
  ];
  it.each(treatmentFilters)(
    "Gets matching foilings for %s",
    (search, expectedTreatments) => {
      const {
        attributes: { treatments },
      } = getKeywordsAndAppliedFiltersFromText(search as string, cards);
      expect(treatments.length).toEqual(expectedTreatments.length);
      expect(treatments).toMatchSnapshot();

      console.log(JSON.stringify({ treatments, expectedTreatments }));

      for (const expected of expectedTreatments as Treatment[]) {
        expect(treatments.includes(expected)).toBeTruthy();
      }
    }
  );

  const metaFilters = [
    ["s:ros meta:rainbow", [Meta.Rainbow]],
    ["meta:rainbow,expansion", [Meta.Rainbow, Meta.Expansion]],
  ];
  it.each(metaFilters)(
    "Gets matching meta values for %s",
    (search, expectedMetaValues) => {
      const { appliedFilters } = getKeywordsAndAppliedFiltersFromText(
        search as string,
        cards
      );

      const metaAppliedFilter = appliedFilters.find(
        (appliedFilter) =>
          appliedFilter.filterToPropertyMapping.property === "meta"
      );

      expect(metaAppliedFilter.values.length).toEqual(
        expectedMetaValues.length
      );

      for (const expected of expectedMetaValues) {
        expect(
          metaAppliedFilter.values.includes(expected.toLowerCase())
        ).toBeTruthy();
      }
    }
  );
});
