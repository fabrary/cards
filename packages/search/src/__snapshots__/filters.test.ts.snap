// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gets the right attribute filters Gets matching foilings for foil:cf,g 1`] = `
[
  "Cold",
  "Gold",
]
`;

exports[`Gets the right attribute filters Gets matching foilings for foiling:r 1`] = `
[
  "Rainbow",
]
`;

exports[`Gets the right attribute filters Gets matching foilings for treat:"alt art",fa 1`] = `
[
  "Alternate Art",
  "Full Art",
]
`;

exports[`Gets the right attribute filters Gets matching foilings for treatment:ea 1`] = `
[
  "Extended Art",
]
`;

exports[`Gets the right attribute filters Gets matching sets for s:ara 1`] = `
[
  "Arakni Blitz Deck",
]
`;

exports[`Gets the right attribute filters Gets matching sets for s:welcome 1`] = `
[
  "Welcome to Rathe",
  "Ira Welcome Deck",
]
`;
